### PR TITLE
[GEOS-10427] Improve access check in ImportProcess

### PR DESCRIPTION
**Pull request by sikeoka**
[GEOS-10427](https://osgeo-org.atlassian.net/browse/GEOS-10427)

This PR adds a check to ensure that users executing the Import process have sufficient privileges to modify the target workspace.

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by aaime** [comment]
Service code should not be mixed with security, it should happen somewhere else... the likely place is the secured catalog wrapper, I'd guess.

**Comment by sikeoka** [comment]
I need more information about what you want.  I was trying to avoid modifying SecureCatalogImpl just to fix an issue in one small part of an extension module.

I suppose I could move the check to org.geoserver.security.SecureCatalogImpl.add(StoreInfo) and org.geoserver.security.SecureCatalogImpl.save(StoreInfo) and see if anything breaks:
https://github.com/geoserver/geoserver/blob/main/src/main/src/main/java/org/geoserver/security/SecureCatalogImpl.java#L1302
https://github.com/geoserver/geoserver/blob/main/src/main/src/main/java/org/geoserver/security/SecureCatalogImpl.java#L1521

There are at least wo other classes that directly access SecureCatalogImpl to check permissions:
https://github.com/geoserver/geoserver/tree/main/src/web/core/src/main/java/org/geoserver/web/WorkspaceAdminComponentAuthorizer.java
https://github.com/geoserver/geoserver/tree/main/src/community/taskmanager/core/src/main/java/org/geoserver/taskmanager/util/TaskManagerSecurityUtil.java

**Comment by aaime** [comment]
The WorkspaceAdminComponentAuthorizer is effectively part of the security subsystem, it's there to make sure the GUI does not list pages/components that the user cannot see (rather than getting exceptions trying to access them). It's properly factored out, no page mocks with security besides declaring (via its base class) if it's an admin page, or not.

The community module is instead breaking the pattern, shouldn't do it. But as a community module, anything goes.

**Comment by aaime** [comment]
Taking a note that this is a good topic for the next PSC meeting, happening tomorrow, 6:30 CET, on Jitsi: https://meet.jit.si/GeoServerMeeting

File: src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/ImportProcess.java [hunk 1]
```diff
@@ -247,6 +254,21 @@ public String execute(
...
+        // This is essentially the same code to check the access level as used by
+        // org.geoserver.security.SecureCatalogImpl.getDataStoreByName(String, String)
+        WrapperPolicy policy =
+                GeoServerExtensions.bean(SecureCatalogImpl.class)
+                        .buildWrapperPolicy(
+                                SecurityContextHolder.getContext().getAuthentication(),
+                                workspace,
+                                workspace.getName(),
+                                MixedModeBehavior.CHALLENGE);
+        if (policy.level != AccessLevel.READ_WRITE) {
```
Comment (aaime) [inline src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/ImportProcess.java:267]: As per our meeting today, moving the checks to the SecureCatalogImpl would be architecturally better, but would open a can of works vs [the REST security](https://docs.geoserver.org/maintain/en/user/security/rest.html).

So ok to do it here, but it should check whether the current user has admin access on the current workspace, something like ``((WorkspaceAccessLimits) policy.getAccessLimits()).isAdminable()``.
Which is probably similar, in practice, to what you're doing, but clearer and we hope more future proof.

Comment (sikeoka) [reply src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/ImportProcess.java]: I updated it to check for admin access instead of write access.

**Comment by aaime** [comment]
Merged. Backport as you see fit.

**Pull request merged**